internal/credential/vault/private_library.go

```diff
@@ -0,0 +1,246 @@
+		}
+		ptkv := &ptk{
+			CtToken: pl.CtToken,
+		}
+		if err := structwrapping.UnwrapStruct(ctx, cipher, ptkv, nil); err != nil {
+			return errors.Wrap(err, op, errors.WithCode(errors.Decrypt), errors.WithMsg("token"))
+		}
+		pl.Token = ptkv.Token
+	}
+
+	if pl.CtClientKey != nil && pl.ClientCert != nil {
+		type pck struct {
+			Key   []byte `wrapping:"pt,key_data"`
+			CtKey []byte `wrapping:"ct,key_data"`
+		}
+		pckv := &pck{
+			CtKey: pl.CtClientKey,
+		}
+		if err := structwrapping.UnwrapStruct(ctx, cipher, pckv, nil); err != nil {
+			return errors.Wrap(err, op, errors.WithCode(errors.Decrypt), errors.WithMsg("client certificate"))
+		}
+		pl.ClientKey = pckv.Key
+	}
+	return nil
+}
+
+func (pl *privateLibrary) client() (*client, error) {
+	const op = "vault.(privateLibrary).client"
+	clientConfig := &clientConfig{
+		Addr:          pl.VaultAddress,
+		Token:         string(pl.Token),
+		CaCert:        pl.CaCert,
+		TlsServerName: pl.TlsServerName,
+		TlsSkipVerify: pl.TlsSkipVerify,
+		Namespace:     pl.Namespace,
+	}
+
+	if pl.ClientKey != nil {
+		clientConfig.ClientCert = pl.ClientCert
+		clientConfig.ClientKey = pl.ClientKey
+	}
+
+	client, err := newClient(clientConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, op, errors.WithMsg("unable to create vault client"))
+	}
+	return client, nil
+}
+
+// TableName returns the table name for gorm.
+func (pl *privateLibrary) TableName() string {
+	return "credential_vault_library_private"
+}
+
+func (r *Repository) getPrivateLibraries(ctx context.Context, libIds []string) ([]*privateLibrary, error) {
+	const op = "vault.(Repository).getPrivateLibraries"
+	if len(libIds) == 0 {
+		return nil, errors.New(errors.InvalidParameter, op, "no library ids")
+	}
+
+	inClause := strings.TrimSuffix(strings.Repeat("?,", len(libIds)), ",")
+
+	query := fmt.Sprintf(selectPrivateLibrariesQuery, inClause)
+
+	var params []interface{}
+	for _, v := range libIds {
+		params = append(params, v)
+	}
+	rows, err := r.reader.Query(ctx, query, params)
+	if err != nil {
+		return nil, errors.Wrap(err, op, errors.WithMsg("query failed"))
+	}
+	defer rows.Close()
+
+	var libs []*privateLibrary
+	for rows.Next() {
+		var lib privateLibrary
+		if err := r.reader.ScanRows(rows, &lib); err != nil {
+			return nil, errors.Wrap(err, op, errors.WithMsg("scan row failed"))
+		}
+		libs = append(libs, &lib)
+	}
+
+	for _, pl := range libs {
+		databaseWrapper, err := r.kms.GetWrapper(ctx, pl.ScopeId, kms.KeyPurposeDatabase)
+		if err != nil {
+			return nil, errors.Wrap(err, op, errors.WithMsg("unable to get database wrapper"))
+		}
+
+		if err := pl.decrypt(ctx, databaseWrapper); err != nil {
+			return nil, errors.Wrap(err, op)
+		}
+	}
+
+	return libs, nil
+}
+
+type privPurpLibrary struct {
+	*privateLibrary
+	credential.Purpose
+}
+
+func allocPrivPurpLibrary() *privPurpLibrary {
+	return &privPurpLibrary{
+		privateLibrary: &privateLibrary{},
+	}
+}
+
+type requestMap struct {
```

Comment:
The purpose of this isn't obvious to me.  Can you add a comment explaining what it is and how it is used?

Reply:
Done. Please let me know if that clears it up.